### PR TITLE
Add API to create a session from model memory buffer with external data path

### PR DESCRIPTION
**Pull request by HectorSVC**
Add new API to create a session from model memory buffer with external data path

### Description
Given a model with external weights. There's no way to load the external weight if create the session using API CreateSessionFromArray. Because it lost track of the model path.
User want to create the session from model memory buffer, and at the mean time provide the path to the external data.
New API required to meet the requirement.

### CODE REVIEW

File: include/onnxruntime/core/graph/graph.h [hunk 1]
```diff
@@ -679,6 +679,9 @@ class Graph {  // NOLINT(clang-analyzer-optin.performance.Padding): preserve exi
   /** Gets the path of the owning model, if any. */
   const Path& ModelPath() const;
 
+  /** Gets the path of the external initializer data, if any. */
+  const Path& ExternalIniPath() const;
```
Comment (skottmckay) [inline include/onnxruntime/core/graph/graph.h:683]: nit: 'ini' is pretty vague, especially when used deep in the code. would ExternalDataPath be clearer (and that's also the param name in the new API?

**Comment by skottmckay** [comment]
I don't quite understand the user requirement. If the weights are in a directory, what is the benefit from loading the model using a buffer?

**Comment by HectorSVC** [comment]
> I don't quite understand the user requirement. If the weights are in a directory, what is the benefit from loading the model using a buffer?

One scenario is the model is loaded into memory once and share it across sessions for different EPs. Maybe common for LLM models.

**Comment by skottmckay** [comment]
> > I don't quite understand the user requirement. If the weights are in a directory, what is the benefit from loading the model using a buffer?
> 
> One scenario is the model is loaded into memory once and share it across sessions for different EPs. Maybe common for LLM models.

In that scenario I would have expected the weights to also be loaded into memory once and shared. When the weights are external the model size should be small, hence I would have expected the largest memory saving would come from sharing the weights not the model.

If the weights are loaded that way is a path to the external weights required?

**Comment by HectorSVC** [comment]
>         model_path.IsEmpty() ? nullptr : model_path.ParentPath().ToPathString().c_str(),

May need to change to use external_data_path.

---
Refers to: onnxruntime/core/framework/tensorprotoutils.cc:611 in bd1e22d. [](commit_id = bd1e22d65dc49fd5bd26fa3411a6872f82105570, deletion_comment = False)

**Pull request closed without merging**